blingfire/src/lib.rs

```diff
@@ -1,100 +1,71 @@
 mod errors;
 
 use blingfire_sys::{TextToSentences as text_to_sentences_ffi, TextToWords as text_to_words_ffi};
-use failchain::{bail, ResultExt};
-use std::{convert::TryInto, mem};
+use failchain::ensure;
+use std::convert::TryInto;
+use std::i32;
+use std::os::raw::{c_char, c_int};
 
 pub use crate::errors::{ErrorKind, Result};
 
 #[inline]
 pub fn text_to_words(source: &str, destination: &mut String) -> Result<()> {
-    tokenize(
-        |raw_source, raw_source_len, raw_destination, raw_destination_capacity| unsafe {
-            text_to_words_ffi(
-                raw_source,
-                raw_source_len,
-                raw_destination,
-                raw_destination_capacity,
-            )
-        },
-        source,
-        destination,
-    )
+    tokenize(text_to_words_ffi, source, destination)
 }
 
 #[inline]
 pub fn text_to_sentences(source: &str, destination: &mut String) -> Result<()> {
-    tokenize(
-        |raw_source, raw_source_len, raw_destination, raw_destination_capacity| unsafe {
-            text_to_sentences_ffi(
-                raw_source,
-                raw_source_len,
-                raw_destination,
-                raw_destination_capacity,
-            )
-        },
-        source,
-        destination,
-    )
+    tokenize(text_to_sentences_ffi, source, destination)
 }
 
+type Tokenizer = unsafe extern "C" fn(*const c_char, c_int, *mut c_char, c_int) -> c_int;
+
+// The maximum length is when the destination length can't fit in the maximum blingfire array. Worst
+// case, the source is tokenized by adding a space after every character + a null character at the
+// end.
+//
+// The maximum array size is defined in blingfire-sys/BlingFire/blingfireclient.library/inc/FALimits.h
+const MAX_SOURCE_LENGTH: usize = 1_000_000_000 / 2 - 2;
```

Comment:
Cool :+1:  if they define a maximum size -- what I had before was just if the capacity of destination was bigger than a `i32`). You should reference this constant from the blingfire library, not duplicate the value here though (whitelist the constant in `generate_bindings`)

Reply:
The constant isn't in the library we're linking against, I don't think this is trivial to expose. Also what you were doing before wasn't quite right, both because the destination **can** be larger, it doesn't hurt anything as long as the source is sufficiently small such that tokenized if fits in the destination and because the C function could never tell you that the destination should larger than `i32::MAX` (since it returns an `i32`).

Reply:
It is part of the library, I tried it out. You just have to add `--whitelist-type FALimits` to `generate_bindings`, it will create a bunch of constants:
```
pub const FALimits_MaxArrSize: FALimits__bindgen_ty_1 = 1000000000;
pub const FALimits_MaxChainSize: FALimits__bindgen_ty_1 = 1000000000;
pub const FALimits_MaxStateVal: FALimits__bindgen_ty_1 = 1000000000;
pub const FALimits_MaxIwVal: FALimits__bindgen_ty_1 = 1000000000;
pub const FALimits_MaxIw: FALimits__bindgen_ty_1 = 1000000000;
pub const FALimits_MaxLdbDumpCount: FALimits__bindgen_ty_1 = 105;
pub const FALimits_MaxWordLen: FALimits__bindgen_ty_1 = 300;
pub const FALimits_MaxWordSize: FALimits__bindgen_ty_1 = 300;
pub const FALimits_MaxWordCount: FALimits__bindgen_ty_1 = 1000;
pub const FALimits_MaxParadigmSize: FALimits__bindgen_ty_1 = 1000;
pub const FALimits_MaxTagsPerWord: FALimits__bindgen_ty_1 = 100;
pub const FALimits_MinTag: FALimits__bindgen_ty_1 = 1;
pub const FALimits_MaxTag: FALimits__bindgen_ty_1 = 65535;
pub const FALimits_MaxGram: FALimits__bindgen_ty_1 = 4;
```

Reply:
TBH, probs useful to export from the library too, stuff like max word len etc. can be useful